source/MatchmakingGraphRoot.cpp

```diff
@@ -0,0 +1,178 @@
+	screenHeight = dimen.height;
+	// Initialize the scene to a locked width
+	if (assets == nullptr) {
+		return false;
+	} else if (!Scene::init(dimen)) {
+		return false;
+	}
+
+	// Start up the input handler
+	this->assets = assets;
+
+	// Acquire the scene built by the asset loader and resize it the scene
+	auto scene = assets->get<Node>("matchmaking");
+	scene->setContentSize(dimen);
+	scene->doLayout(); // Repositions the HUD
+
+	// Get the scene components.
+	host = std::dynamic_pointer_cast<Button>(assets->get<Node>("matchmaking_host"));
+	client = std::dynamic_pointer_cast<Button>(assets->get<Node>("matchmaking_client"));
+	roomLabel = std::dynamic_pointer_cast<Label>(assets->get<Node>("matchmaking_room"));
+	roomInput = std::dynamic_pointer_cast<TextField>(assets->get<Node>("matchmaking_input"));
+	textInput = std::dynamic_pointer_cast<Button>(assets->get<Node>("matchmaking_inputbutton"));
+
+	addChild(scene);
+	return true;
+}
+
+/**
+ * Disposes of all (non-static) resources allocated to this mode.
+ */
+void MatchmakingGraphRoot::dispose() {
+	if (_active) {
+		removeAllChildren();
+		host = nullptr;
+		client = nullptr;
+		roomInput = nullptr;
+		roomLabel = nullptr;
+		_active = false;
+	}
+}
+
+#pragma mark -
+#pragma mark Gameplay Handling
+
+/**
+ * Resets the status of the game so that we can play again.
+ */
+void MatchmakingGraphRoot::reset() {}
+
+/**
+ * The method called to update the game mode.
+ *
+ * This method contains any gameplay code that is not an OpenGL call.
+ *
+ * @param timestep  The amount of time (in seconds) since the last frame
+ */
+void MatchmakingGraphRoot::update(float timestep) {
+	// "Drawing" code.  Move everything BUT the donut
+	// Update the HUD
+	roomLabel->setText(positionText());
+}
+
+/**
+ * Returns integers representing which button has been tapped if any
+ *
+ * @param position The screen coordinates of the tap
+ *
+ * @return -1 if no button pressed 0 for host creation, 1 for client creation
+ */
+int MatchmakingGraphRoot::checkButtons(const cugl::Vec2& position) {
+	if (position == Vec2::ZERO) {
+		return -1;
+	} else if (host->containsScreen(position)) {
+		playerId = 0;
+		roomLabel->setVisible(true);
+		host->setVisible(false);
+		client->setVisible(false);
+		return 0;
+	} else if (client->containsScreen(position)) {
+		roomInput->setVisible(true);
+		roomInput->activate(2);
+		roomLabel->setVisible(true);
+		textInput->setVisible(true);
+		return 1;
+	} else {
+		return -1;
+	}
+}
+
+/**
+ * Callback for TextInput
+ *
+ * @param event The text input event
+ *
+ */
+/*void MatchmakingGraphRoot::textInputEnded(const cugl::TextInputEvent& event, bool focus) {
+	roomId = event.buffer;
+	CULog("Text Input Callback");
+}*/
+
+/**
+ * Returns the text from inside the input field
+ *
+ * @return a string for the text
+ */
+std::string MatchmakingGraphRoot::getInput(const cugl::Vec2& position) {
+	if (position == Vec2::ZERO) {
+		return "";
```

Comment:
I'm confused why this is here

Reply:
The position vector is  a vec:zero if no location has been tapped, so I just put a catch in here

Reply:
Sure, but why use zero instead of something like null? What if the user actually tapped on `(0,0)`?

Reply:
Null is defined as 0 somewhere and can’t be directly compared iirc and nullptrs can’t be compared to vec2 cus they’re not pointers

Reply:
Null is defined as 0 somewhere and can’t be directly compared iirc and nullptrs can’t be compared to vec2 cus they’re not pointers

Reply:
As for tapping on 0,0 I find it incredibly unlikely they tap on a very specific pixel when the location is a vec2 of floats

Reply:
We should probably better abstract this out in `InputController` later, then. We'll let this slide for now.

Reply:
*Specific coordinate